### PR TITLE
NIFI-12959: Support loading python processors from NARs

### PR DESCRIPTION
**Pull request by markap14**
# Summary

[NIFI-00000](https://issues.apache.org/jira/browse/NIFI-00000)

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [ ] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [ ] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [ ] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [ ] Pull Request based on current revision of the `main` branch
- [ ] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [ ] Build completed using `mvn clean install -P contrib-check`
  - [ ] JDK 21

### Licensing

- [ ] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [ ] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [ ] Documentation formatting appears as expected in rendered files

### CODE REVIEW

File: nifi-nar-bundles/nifi-py4j-bundle/nifi-py4j-bridge/src/test/java/org/apache/nifi/py4j/PythonProcessTest.java [hunk 1]
```diff
@@ -56,51 +56,46 @@ class PythonProcessTest {
...
-        final String expected = getExpectedBinaryPath(WINDOWS_SCRIPTS_DIR);
-        assertEquals(expected, result);
+        final String expected = getExpectedBinaryPath(WINDOWS_SCRIPTS_DIR); assertEquals(expected, result);
     }
 
     @Test
     void testResolvePythonCommandUnix() throws IOException {
-        final File binDir = new File(virtualEnvHome, UNIX_BIN_DIR);
-        assertTrue(binDir.mkdir());
+        final File binDir = new File(virtualEnvHome, UNIX_BIN_DIR); assertTrue(binDir.mkdir());
```
Comment (dan-s1) [inline nifi-nar-bundles/nifi-py4j-bundle/nifi-py4j-bridge/src/test/java/org/apache/nifi/py4j/PythonProcessTest.java:80]:
```suggestion
        final File binDir = new File(virtualEnvHome, UNIX_BIN_DIR);
        assertTrue(binDir.mkdir());

```

File: nifi-nar-bundles/nifi-py4j-bundle/nifi-py4j-bridge/src/test/java/org/apache/nifi/py4j/PythonProcessTest.java [hunk 2]
```diff
@@ -56,51 +56,46 @@ class PythonProcessTest {
...
 
-        when(pythonProcessConfig.getPythonCommand()).thenReturn(PYTHON_CMD);
-        final String result = this.pythonProcess.resolvePythonCommand();
+        when(pythonProcessConfig.getPythonCommand()).thenReturn(PYTHON_CMD); final String result = this.pythonProcess.resolvePythonCommand();
 
-        final String expected = getExpectedBinaryPath(UNIX_BIN_DIR);
-        assertEquals(expected, result);
+        final String expected = getExpectedBinaryPath(UNIX_BIN_DIR); assertEquals(expected, result);
```
Comment (dan-s1) [inline nifi-nar-bundles/nifi-py4j-bundle/nifi-py4j-bridge/src/test/java/org/apache/nifi/py4j/PythonProcessTest.java:84]:
```suggestion
        final String expected = getExpectedBinaryPath(UNIX_BIN_DIR);
        assertEquals(expected, result);

```

File: nifi-nar-bundles/nifi-py4j-bundle/nifi-py4j-bridge/src/test/java/org/apache/nifi/py4j/PythonProcessTest.java [hunk 3]
```diff
@@ -56,51 +56,46 @@ class PythonProcessTest {
...
     }
 
     @Test
     void testResolvePythonCommandPreferBin() throws IOException {
-        final File binDir = new File(virtualEnvHome, UNIX_BIN_DIR);
-        assertTrue(binDir.mkdir());
-        final File scriptsDir = new File(virtualEnvHome, WINDOWS_SCRIPTS_DIR);
-        assertTrue(scriptsDir.mkdir());
+        final File binDir = new File(virtualEnvHome, UNIX_BIN_DIR); assertTrue(binDir.mkdir()); final File scriptsDir = new File(virtualEnvHome, WINDOWS_SCRIPTS_DIR); assertTrue(scriptsDir.mkdir());
```
Comment (dan-s1) [inline nifi-nar-bundles/nifi-py4j-bundle/nifi-py4j-bridge/src/test/java/org/apache/nifi/py4j/PythonProcessTest.java:89]:
```suggestion
        final File binDir = new File(virtualEnvHome, UNIX_BIN_DIR);
        assertTrue(binDir.mkdir());
        final File scriptsDir = new File(virtualEnvHome, WINDOWS_SCRIPTS_DIR);
        assertTrue(scriptsDir.mkdir());

```

File: nifi-nar-bundles/nifi-py4j-bundle/nifi-py4j-bridge/src/test/java/org/apache/nifi/py4j/PythonProcessTest.java [hunk 4]
```diff
@@ -56,51 +56,46 @@ class PythonProcessTest {
...
 
-        when(pythonProcessConfig.getPythonCommand()).thenReturn(PYTHON_CMD);
-        final String result = this.pythonProcess.resolvePythonCommand();
+        when(pythonProcessConfig.getPythonCommand()).thenReturn(PYTHON_CMD); final String result = this.pythonProcess.resolvePythonCommand();
```
Comment (dan-s1) [inline nifi-nar-bundles/nifi-py4j-bundle/nifi-py4j-bridge/src/test/java/org/apache/nifi/py4j/PythonProcessTest.java:91]:
```suggestion
        when(pythonProcessConfig.getPythonCommand()).thenReturn(PYTHON_CMD);
        final String result = this.pythonProcess.resolvePythonCommand();

```

File: nifi-nar-bundles/nifi-py4j-bundle/nifi-py4j-bridge/src/test/java/org/apache/nifi/py4j/PythonProcessTest.java [hunk 5]
```diff
@@ -56,51 +56,46 @@ class PythonProcessTest {
...
 
-        final String expected = getExpectedBinaryPath(UNIX_BIN_DIR);
-        assertEquals(expected, result);
+        final String expected = getExpectedBinaryPath(UNIX_BIN_DIR); assertEquals(expected, result);
```
Comment (dan-s1) [inline nifi-nar-bundles/nifi-py4j-bundle/nifi-py4j-bridge/src/test/java/org/apache/nifi/py4j/PythonProcessTest.java:93]:
```suggestion
        final String expected = getExpectedBinaryPath(UNIX_BIN_DIR);
        assertEquals(expected, result);

```

File: nifi-nar-bundles/nifi-py4j-bundle/nifi-py4j-bridge/src/test/java/org/apache/nifi/py4j/PythonProcessTest.java [hunk 6]
```diff
@@ -56,51 +56,46 @@ class PythonProcessTest {
...
     }
 
     @Test
     void testResolvePythonCommandNone() {
-        when(pythonProcessConfig.getPythonCommand()).thenReturn(PYTHON_CMD);
-        assertThrows(IOException.class, ()-> this.pythonProcess.resolvePythonCommand());
+        when(pythonProcessConfig.getPythonCommand()).thenReturn(PYTHON_CMD); assertThrows(IOException.class, () -> this.pythonProcess.resolvePythonCommand());
```
Comment (dan-s1) [inline nifi-nar-bundles/nifi-py4j-bundle/nifi-py4j-bridge/src/test/java/org/apache/nifi/py4j/PythonProcessTest.java:98]:
```suggestion
        when(pythonProcessConfig.getPythonCommand()).thenReturn(PYTHON_CMD);
        assertThrows(IOException.class, () -> this.pythonProcess.resolvePythonCommand());

```

**Comment by pvillard31** [comment]
Nice improvement Mark! Would this approach help with a processor like ParseDocument where Tesseract has to be installed on the node first?

**Comment by markap14** [comment]
@pvillard31 no, it wouldn't really help there. The issue there is that a native library like Tesseract would need to be compiled for the correct OS and architecture. So in order to include that, we'd need to include many copies of the library. And currently, we do not set any sort of environment variables telling it to search for native libraries in the NAR's unpacked directory. It might be worth exploring as a future improvement, though. Even though it may not help for something like ParseDocument, it might be helpful at least for custom nars, where perhaps you know that you're only going to run in containers or on a specific architecture/os so you can include the appropriate library for the custom nar?

**Comment by pvillard31** [comment]
Yeah, I must admit that I tried to figure out a way to use ParseDocument with NiFi in a container and I didn't manage to find a clean way to bring Tesseract into it. I feel like it would probably be easier to have it as a side container or something but I didn't find a way to piece everything together so I was naively (without too much hopes) thinking that this improvement may help :)

**Comment by markap14** [comment]
@dan-s1 thanks great catch. I removed some carrige-return-newline combos in that file and my IDE did something I didn't expect :) Fixed that.

File: nifi-system-tests/nifi-system-test-suite/src/test/java/org/apache/nifi/tests/system/python/PythonNarIT.java [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+            deleteRecursively(extensionFile);
+        }
+
+        // Restart NiFi
+        getNiFiInstance().stop();
+        getNiFiInstance().start(true);
+
+        // Create instance of the WriteNumpyVersionBundledDependency processor, and connect a GenerateFlowFile to it
+        final ProcessorEntity generate = getClientUtil().createProcessor("GenerateFlowFile");
+        ProcessorEntity writeVersion = getClientUtil().createPythonProcessor("WriteJson5VersionBundledDependency");
```
Comment (exceptionfactory) [inline nifi-system-tests/nifi-system-test-suite/src/test/java/org/apache/nifi/tests/system/python/PythonNarIT.java:75]: Note that this reference should be changed to match the updated Processor name.

File: nifi-system-tests/nifi-system-test-suite/src/test/java/org/apache/nifi/tests/system/python/PythonNarIT.java [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+
+        // Run the flow
+        getClientUtil().startProcessor(generate);
+        waitForQueueCount(generateToWriteVersion.getId(), 1);
+        getClientUtil().startProcessor(writeVersion);
+        waitForQueueCount(writeVersionToTerminate.getId(), 1);
+
+        // Verify the output
+        final String contents = getClientUtil().getFlowFileContentAsUtf8(writeVersionToTerminate.getId(), 0);
+        assertEquals("qpzry9x8gf2tvdw0s3jn54khce6mua7l", contents);
```
Comment (exceptionfactory) [inline nifi-system-tests/nifi-system-test-suite/src/test/java/org/apache/nifi/tests/system/python/PythonNarIT.java:95]: Recommend calling this string out as a static variable named `BECH32_CHARSET`.

**Review by exceptionfactory — CHANGES_REQUESTED** [review]
Thanks for the improvement @markap14, this will be helpful for Python Processor deployment! I just noted a name change to reflect the Python library used for testing.

File: nifi-nar-bundles/nifi-py4j-bundle/nifi-py4j-bridge/src/main/java/org/apache/nifi/py4j/PythonProcess.java [hunk 9]
```diff
@@ -287,6 +309,13 @@ String resolvePythonCommand() throws IOException {
 
 
     private void setupEnvironment() throws IOException {
+        // Environment creation is only necessary if using PIP. Otherwise, the Process requires no outside dependencies, other than those
+        // provided in the package and thus we can simply include those packages in the PYTHON_PATH.
+        if (!isUseVirtualEnv()) {
+            logger.debug("Will not create Python Virtual Environment because PIP is disabled in nifi.properties");
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-py4j-bundle/nifi-py4j-bridge/src/main/java/org/apache/nifi/py4j/PythonProcess.java:315]: Very minor, but general usage of `pip` is all lowercase and opposed to all uppercase. More importantly though, the message does not reflect the behavior, as it is only based on the the presence of packaged dependencies.
```suggestion
            logger.debug("Python Virtual Environment not created: Python Processor packaged with dependencies");
```

File: nifi-nar-bundles/nifi-py4j-bundle/nifi-py4j-bridge/src/main/java/org/apache/nifi/py4j/PythonProcess.java [hunk 10]
```diff
@@ -224,6 +226,10 @@ private String generateAuthToken() {
         return Base64.getEncoder().encodeToString(bytes);
     }
 
+    private boolean isUseVirtualEnv() {
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-py4j-bundle/nifi-py4j-bridge/src/main/java/org/apache/nifi/py4j/PythonProcess.java:229]: It looks like the only use of this method is a negated reference. Can this be removed and replaced with `if (packagedWithDependencies)`?

File: nifi-nar-bundles/nifi-py4j-bundle/nifi-py4j-bridge/src/main/java/org/apache/nifi/py4j/PythonProcess.java [hunk 11]
```diff
@@ -234,8 +240,18 @@ private Process launchPythonProcess(final int listeningPort, final String authTo
...
+        if (!isUseVirtualEnv()) {
+            // If not using venv, we will not launch a separate virtual environment, so we need to use the -S
+            // flag in order to prevent the Python process from using the installation's site-packages. This provides
+            // proper dependency isolation to the Python process.
+            commands.add("-S");
+        }
 
         String pythonPath = pythonApiDirectory.getAbsolutePath();
+        final String absolutePath = virtualEnvHome.getAbsolutePath();
+        final File dependenciesDir = new File(new File(absolutePath), "NAR-INF/bundled-dependencies");
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-py4j-bundle/nifi-py4j-bridge/src/main/java/org/apache/nifi/py4j/PythonProcess.java:252]: Is the intent to add `NAR-INF/bundled-dependencies` in all cases? It seems like it should be conditional on whether the Processor is packaged with dependencies.

**Review by exceptionfactory — COMMENTED** [review]
Thanks for the updates @markap14, I noted a couple more minor questions, otherwise this looks ready to go.

File: nifi-nar-bundles/nifi-py4j-bundle/nifi-py4j-bridge/src/main/java/org/apache/nifi/py4j/PythonProcess.java (hunk 11 above)
Comment (markap14) [reply nifi-nar-bundles/nifi-py4j-bundle/nifi-py4j-bridge/src/main/java/org/apache/nifi/py4j/PythonProcess.java:252]: Good catch.

**Review by exceptionfactory — APPROVED** [review]
Thanks for making the adjustments @markap14, the latest version looks good! +1 merging

**Comment by samer1977** [comment]
Hi ,

When I try to follow the folder structure:
```

my-nar.nar
+-- META-INF/
    +-- MANIFEST.MF
+-- NAR-INF/
    +-- bundled-dependencies/
        +-- dependency1
        +-- dependency2
        +-- etc.
+-- MyProcessor.py
```
I keep getting the following error: 
`java.io.FileNotFoundException: work\nar\extensions\nifi-MyProcessor-nar-2.0.0-M4.nar-unpacked\MyProcessor.py (The system cannot find the path specified)`

Please advise.

**Pull request closed without merging**